lib/vanagon/project.rb

```diff
@@ -270,22 +270,75 @@ def has_provides?
       !get_provides.empty?
     end
 
-    # Collects the preinstall packaging actions for the project and it's components
-    # for the specified packaging state
+    # Collects the preinstall packaging actions for the project and its components
+    #   for the specified packaging state
     #
     # @param pkg_state [String] the package state we want to run the given scripts for.
     #   Can be one or more of 'install' or 'upgrade'
     # @return [String] string of Bourne shell compatible scriptlets to execute during the preinstall
     #   phase of packaging during the state of the system defined by pkg_state (either install or upgrade)
-    def get_preinstall_actions(pkg_state)
-      scripts = @components.map(&:preinstall_actions).flatten.compact.select { |s| s.pkg_state.include? pkg_state }.map(&:scripts)
-      if scripts.empty?
+    def get_preinstall_actions(pkg_state) # rubocop:disable Metrics/AbcSize
+      pkg_state = Array(pkg_state)
+      preinstall_scripts = Array.new
+      if pkg_state.empty? || !(pkg_state - ["install", "upgrade"]).empty?
+        raise Vanagon::Error, "#{pkg_state} should be an array containing one or more of ['install', 'upgrade']"
+      end
+      preinstalls = @components.map(&:preinstall_actions).flatten.compact.select { |s| s.pkg_state == pkg_state }
```

Comment:
Will `components.flat_map(&:preinstall_actions).select { |s| s.pkg_state == pkg_state }` work here?

(If you're calling `#compact`, `#flatten`, and `#select`, that's a light code smell. I don't know what your data structure looks like offhand, so I might be wrong. Test it (if you can do it quickly) and see?

Reply:
Looks like it works with the change!

Reply:
I'm kind of goofing around here, but does this method work if rewritten as follows? It passes tests but you're closer to this problem than I am.

```ruby
def get_preinstall_actions(pkg_state) # rubocop:disable Metrics/AbcSize
  pkg_state = Array(pkg_state)

  if pkg_state.empty? || (pkg_state & %w[install upgrade]).any?
    raise Vanagon::Error,
          "#{pkg_state} should be an array containing one or more of ['install', 'upgrade']"
  end

  preinstall_scripts = components.each_with_object([]) do |component, arr|
    arr.push(component.preinstall_actions.select { |action| action.pkg_state == pkg_state })
  end

  return ': no preinstall scripts provided' if preinstall_scripts.empty?
  preinstall_scripts.flatten.compact.join("\n")
end
```

I tried to get the ABC size down (I got it to 16.4 by reducing intermediate assignment and 15.33 by removing coercion; was aiming for a straight 15 but c'est la vie), but I may have rewritten it too much. Anyway, I wanted to introduce you to Ruby's [`Enumerable#each_with_object`](https://ruby-doc.org/core-2.2.0/Enumerable.html) method. If I wrote this right, then this will allow you to cut down on passes over the first-level Enumerable object (hash, array, set, etc.) that is `#components`/`@components`.

Reply:
Oh. I can get the ABC score down to/below 15 if I yank the `#compact` and `#flatten` off the final return. Comment updated. Still not super-sure this works right -- the test case doesn't seem to completely exercise this method.